### DISCUSSION TITLE
Event dispatch for restarting checks on demand

### DISCUSSION THREAD

**Discussion by lppedd** [quarkiverse/quarkus-github-app#383] [General]
I have implemented a very simple command:
```
@bot checks restart
```

which is useful when a user wants to restart a set of checks.  
The command implementation is the following:
```kotlin
@Command(
  name = "restart",
  description = "Restart all the Checks for a pull request",
)
@CommandOptions(
  scope = CommandScope.PULL_REQUESTS,
  reactionStrategy = ReactionStrategy.ON_PROGRESS,
)
class ChecksRestartCommand : AppCommand {
  @Inject
  lateinit var event: Event<ChecksRestartRequest>

  override fun run(payload: GHEventPayload.IssueComment): Message {
    val repository = payload.repository
    val issue = payload.issue
    val pullRequest = repository.getPullRequest(issue.number)

    if (!pullRequest.isClosed) {
      event.fireAsync(ChecksRestartRequest(pullRequest))
    }

    return NoMessage
  }
}
```

A possible observer could be:
```kotlin
fun onChecksRestartRequest(@ObservesAsync request: ChecksRestartRequest): Unit =
    checkPullRequest(request.pullRequest)
```

Now, it's all good, it works. But, is there a way to make it more elegant like it's done in this extension?  
I see on the README diagram that CDI events used, but how is the `@Observes*` annotation masked?

An end result I'd appreciate could be:
```kotlin
fun onChecksRestartRequest(@Restart request: ChecksRestartRequest): Unit =
    checkPullRequest(request.pullRequest)
```

**Comment by lppedd** [comment]
Btw, this is a common use-case probably. Wanting to `checks restart all` or `checks restart check-id`.  
Sometime checks hang, or you need to execute one even if it's passing, and you cannot do that through GitHub.

Would it make sense to include a facility in the extension itself?

**Comment by gsmet** [comment]
> But, is there a way to make it more elegant like it's done in this extension?

The way the extension makes it cute uses code generation and this is doable because it's an extension. A Quarkus extension contains some build time components and can be used to alter the build and generate code.

I think the way you do it is good enough and I wouldn't overthink it. It's worth going the extension way when you're building something more general purpose and it's actually a good idea to streamline things for a lot of use cases.

> Btw, this is a common use-case probably. Wanting to checks restart all or checks restart check-id.

I agree on that. But while the feature might be common, the way people will interact with each bot is actually quite specific. For our build reporting stuff, I actually created a library with the feature. Feature that can be reused in different GitHub Apps or even GitHub Actions. See https://github.com/quarkusio/build-reporter .

It might be a good idea at some point to create some sort of library of reusable components.

**Comment by lppedd** [comment]
@gsmet thanks!

> some sort of library of reusable components

Can this be applied to commands too? I'd really like to extract some of them to a common library.

**Reply by gsmet** [reply]
Your commands can be in an external library (but the library needs to be indexed by Jandex).

Extracting the commands might not be completely safe for reuse though as you might expose some things specific to your CLI.

Another option is to use Airline composition with `@AirlineModule`.

And of course the last option is to just have a library containing the code that is executed by the commands but not the commands themselves. That's the way I went for the build reporter, also because it needed to be usable both for a GitHub App and a GitHub Action.
